Add `OpenGridCalculation` for `open_grid.x` code

qiaojunfeng opened the pull request:
QE has an `open_grid.x` code to unfold an irreducible-Brillouin-zone calculation to a full kmesh, significantly reduces the computation if the following calculation requires a full kmesh. This PR adds a `OpengridCalculation` class to support the code.

To allow subsequent `ProjwfcCalculation` on top of the `OpengridCalculation`, I have to introduce a somewhat ugly input/output spec:
1. a `structure` input which isn't used by the `OpengridCalculation`, but because `ProjwfcCalculation` implicitly requires the parent calculation having an input `structure` (so it can parse the atomic orbitals).
2. I modified the `ProjwfcParser` to allow it to use output `kpoints` if no input `kpoints` in the parent calculation (projwfc should use the output `kpoints` which is the full kmesh).

sphuber commented:
Thanks @yunfeng. Before I review this, a quick couple of questions:

1. Where exactly is the computation time won? If you pass an IBZ kpoints to a calculation, will QE try to do the unfolding itself but in a less effective way compared to `open_grid.x`? Surely calculating a full mesh is more expensive compared to a reduced mesh?
2. The issue you describe of the `ProjwfcParser` already cropped up a long time before and is documented in #299 . Ideally, we should really just fix this, and then you also don't have to work around this with weird specs etc.

qiaojunfeng commented:
Thanks @sphuber, 

> Thanks @Yunfeng. Before I review this, a quick couple of questions:
> 
> 1. Where exactly is the computation time won? If you pass an IBZ kpoints to a calculation, will QE try to do the unfolding itself but in a less effective way compared to `open_grid.x`? Surely calculating a full mesh is more expensive compared to a reduced mesh?

In essence, the `open_grid.x` just read the wavefunctions of IBZ kpoints, apply some symmetry operations to generate wavefunction at the remaining kpoints. For `pw.x`, we can do a nscf calculation on full kmesh with restarting from IBZ calculation, but that requires diagonalization of the Kohn Sham equations, and orthogonalization of KS orbitals, this is much slower than reading and rewriting wavefunctions. This difference is something like 30 mins v.s. 3 mins. 
> 2. The issue you describe of the `ProjwfcParser` already cropped up a long time before and is documented in [ProjwfcParser uses input nodes up to 3 links back, and prevents easy testing #299](https://github.com/aiidateam/aiida-quantumespresso/issues/299) . Ideally, we should really just fix this, and then you also don't have to work around this with weird specs etc.

Yes I agree we should fix the `ProjwfcParser`, the question is how to fix it: 
the `ProjwfcParser` requires a `kpoints`, a `structure`, and a `output_parameters` containing `number_of_spin_components`,
https://github.com/aiidateam/aiida-quantumespresso/blob/c49def7a238617439feb75afd0d0cf9fbb681613/aiida_quantumespresso/parsers/projwfc.py#L404
In this opengrid case, the `ProjwfcParser` should use the opengrid output `kpoints`, since it is different from its input `kpoints`, so we can add an optional input `kpoints` to `ProjwfcCalculation`. Similarly, we can add an optional input `structure` for `ProjwfcCalculation`. But for `number_of_spin_components`, adding an optional input `parameters` containing `number_of_spin_components` to `ProjwfcCalculation`, is a bit weird. Also the `projwfc.x` raw output does not contain spin info.
Or, in a totally different way, the XML file in the `parent_folder` contains the needed info to generate `kpoints`, `structure`, and `number_of_spin_components`, should we parse the XML in `ProjwfcParser`?

sphuber commented:
> In essence, the `open_grid.x` just read the wavefunctions of IBZ kpoints, apply some symmetry operations to generate wavefunction at the remaining kpoints. For `pw.x`, we can do a nscf calculation on full kmesh with restarting from IBZ calculation, but that requires diagonalization of the Kohn Sham equations, and orthogonalization of KS orbitals, this is much slower than reading and rewriting wavefunctions. This difference is something like 30 mins v.s. 3 mins.

Thanks a lot. So just to see if I get the use case here: after an NSCF performed on a set of kpoints in the IBZ, instead of doing a new NSCF restart on the full mesh before passing it to `projwfc.x`, we first go through `open_grid.x` (instead of `pw.x`) and then directly go to `projwfc.x`?

> Yes I agree we should fix the `ProjwfcParser`, the question is how to fix it:
> the `ProjwfcParser` requires a `kpoints`, a `structure`, and a `output_parameters` containing `number_of_spin_components`,
> 
> https://github.com/aiidateam/aiida-quantumespresso/blob/c49def7a238617439feb75afd0d0cf9fbb681613/aiida_quantumespresso/parsers/projwfc.py#L404
> 
> In this opengrid case, the `ProjwfcParser` should use the opengrid output `kpoints`, since it is different from its input `kpoints`, so we can add an optional input `kpoints` to `ProjwfcCalculation`. Similarly, we can add an optional input `structure` for `ProjwfcCalculation`. 

I am not sure that adding inputs to `ProjwfcCalculation` is the way to go. The whole problem with the `ProjwfcParser` that we have now is that it is not self-contained and reaches up the provenance graph to get information from preceding calculations. Passing in an output of the parent calculation as an input, is just another version of this same workaround.

> But for `number_of_spin_components`, adding an optional input `parameters` containing `number_of_spin_components` to `ProjwfcCalculation`, is a bit weird. Also the `projwfc.x` raw output does not contain spin info.
> Or, in a totally different way, the XML file in the `parent_folder` contains the needed info to generate `kpoints`, `structure`, and `number_of_spin_components`, should we parse the XML in `ProjwfcParser`?

If the output of `projwfc.x` itself does not contain the necessary information, than we should indeed get it from the XML of the parent folder because this is exactly what `projwfc.x` bases itself on. This would also solve the problem of using a `parent_folder` that may come from different calcjobs. In this way, it should not matter for the `ProjwfcParser` and simply always parses the same XML. That is to say that QE doesn't do something incredibly weird and annoying and have `pw.x` and `open_grid.x` write different output formats... :sweat_smile:

qiaojunfeng commented:
Thanks @mbercx for fixing #722, #741, and #747! 
Now I have removed those strange input spec from `OpengridCalculation`, and added some tests.
Finally, the whole Wannier workchain runs smoothly 😄

sphuber commented:
Thanks @qiaojunfeng . I propose we wait for @mbercx 's PRs to be finished and merged and then we rebase your PR. Then I will give it a final pass and we can merge it.

qiaojunfeng commented on aiida_quantumespresso/calculations/opengrid.py:
```diff
@@ -0,0 +1,39 @@
+    _blocked_keywords = [
+        ('INPUTPP', 'outdir', NamelistsCalculation._OUTPUT_SUBFOLDER),
+        ('INPUTPP', 'prefix', NamelistsCalculation._PREFIX),
+        ('INPUTPP', 'overwrite_prefix', True),
```
Since [that commit](https://gitlab.com/QEF/q-e/-/commit/f0e9e696f53ad36c5c1f7d70369387a5106bf9a5) has been included since QE 7.0, so I assume it is not a problem for aiida-qe now^^

mbercx replied:
Well, actually 😅 

https://github.com/aiidateam/aiida-quantumespresso#compatibility-matrix

mbercx replied:
That said, if you can't think of a reasonably easy workaround to also support QE>=6.6, I think it's better to merge the PR than stand on ceremony.

mbercx commented on src/aiida_quantumespresso/parsers/open_grid.py:
```diff
@@ -0,0 +1,80 @@
+        except OSError:
+            return self.exit(self.exit_codes.ERROR_OUTPUT_STDOUT_READ)
+
+        parsed_data, logs = parse_output_base(out_file, codename='OPEN_GRID')
```
Since the returned `parsed_data` here also returns the QE version, perhaps we can also add a warning to the logs here in case the QE version is < 7.0?

On another note, most of the other `NamelistsCalculation`s also output the version and walltime to an `output_parameters` output. Might be nice to add these here as well for consistency, and in case someone e.g. wants to get the full wall time of a work chain.

qiaojunfeng replied:
For QE < 7.0, the `open_grid.x` wouldn't run at all so probably the parser will output empty results.
But the wall time info is indeed useful, added in https://github.com/aiidateam/aiida-quantumespresso/pull/714/commits/a4689cfe23010a91cff766d16ef7939d13afaffe

mbercx reviewed (changes requested):
Great, thanks @qiaojunfeng! Just had a quick look at the changes, and only have a few comments.

qiaojunfeng commented on src/aiida_quantumespresso/parsers/open_grid.py:
```diff
@@ -0,0 +1,80 @@
+        self.out('kpoints', kpoints)
+
+    @classmethod
+    def parse_kpoints(cls, out_file):
```
Yes, changed to `@staticmethod` in https://github.com/aiidateam/aiida-quantumespresso/pull/714/commits/a4689cfe23010a91cff766d16ef7939d13afaffe

mbercx reviewed (approved):
For me this one is good to go! 🚀 @sphuber did you still want to have another look or can I go ahead and merge?

sphuber commented:
Sure looks good to me! Thanks a lot to both. Just make sure to squash merge and add a nice commit message as per yuzh
